File: src/ts/test/generic.test.mjs
import assert from "node:assert/strict";
import test from "node:test";

import {
  composeMiddleware,
  contextualProviderFrom,
  providerFrom
} from "../dist/generic.js";

test("generic provider keeps concrete SDK ownership in the consumer", async () => {
  const sdk = Object.freeze({ version: "v7", prefix: "sdk-v7:" });
  const provider = providerFrom(async (token) => {
    assert.equal(sdk.version, "v7");
    if (!token.startsWith(sdk.prefix)) throw new Error("rejected");
    return { subject: token.slice(sdk.prefix.length) };
  });

  assert.deepEqual(await provider.verify("sdk-v7:alice"), { subject: "alice" });
});

test("contextual provider is generic over request and context types", async () => {
  const provider = contextualProviderFrom(
    async (request, context) => `${context.tenant}:${request.token}`
  );

  assert.equal(
    await provider.verify({ request: { token: "abc" }, context: { tenant: "t-1" } }),
    "t-1:abc"
  );
});

test("middleware order is selected entirely by the consumer", async () => {
  const events = [];
  const stage = (name) => (next) => async (request) => {
    events.push(`${name}:before`);
    const response = await next(request);
    events.push(`${name}:after`);
    return response;
  };

  const handler = async (request) => {
    events.push("handler");
    return `${request}:ok`;
  };

  const composed = composeMiddleware(
    handler,
    stage("request-id"),
    stage("consumer-auth-v7"),
    stage("tenant-rate-limit")
  );

  assert.equal(await composed("request"), "request:ok");
  assert.deepEqual(events, [
    "request-id:before",
    "consumer-auth-v7:before",
    "tenant-rate-limit:before",
    "handler",
    "tenant-rate-limit:after",
    "consumer-auth-v7:after",
    "request-id:after"
  ]);
});

File: src/elixir/test/generic_test.exs
defmodule OresMiddleware.GenericTest do
  use ExUnit.Case, async: true

  alias OresMiddleware.Generic

  test "consumer provider identity and failure values are preserved" do
    sdk = %{prefix: "v17:"}

    provider =
      Generic.provider_from(fn token ->
        if String.starts_with?(token, sdk.prefix),
          do: {:ok, String.replace_prefix(token, sdk.prefix, "")},
          else: {:error, {:bad_token, token}}
      end)

    assert provider.("v17:alice") == {:ok, "alice"}
    assert provider.("wrong") == {:error, {:bad_token, "wrong"}}
  end

  test "contextual provider is generic over request and context" do
    provider =
      Generic.contextual_provider_from(fn request, context ->
        {context.tenant, request.token}
      end)

    assert provider.(%{request: %{token: "abc"}, context: %{tenant: "t-1"}}) ==
             {"t-1", "abc"}
  end

  test "middleware order and duplicate named stages are preserved" do
    parent = self()

    stage = fn name ->
      fn next ->
        fn request ->
          send(parent, {:before, name})
          response = next.(request)
          send(parent, {:after, name})
          response
        end
      end
    end

    handler = fn request ->
      send(parent, :handler)
      request <> ":ok"
    end

    composed =
      Generic.compose_named(handler, [
        %{name: "auth", middleware: stage.("auth")},
        %{name: "auth", middleware: stage.("auth")},
        %{name: "audit", middleware: stage.("audit")}
      ])

    assert composed.("request") == "request:ok"
    assert_receive {:before, "auth"}
    assert_receive {:before, "auth"}
    assert_receive {:before, "audit"}
    assert_receive :handler
    assert_receive {:after, "audit"}
    assert_receive {:after, "auth"}
    assert_receive {:after, "auth"}
  end

  test "empty chains preserve handler identity" do
    handler = fn request -> request end
    contextual = fn request, context -> {request, context} end
    assert Generic.compose(handler, []) === handler
    assert Generic.compose_named(handler, []) === handler
    assert Generic.compose_contextual(contextual, []) === contextual
    assert Generic.compose_named_contextual(contextual, []) === contextual
  end

  test "contextual middleware preserves caller-owned context" do
    stage = fn next ->
      fn request, context -> next.(request, Map.put(context, :seen, true)) end
    end

    handler = fn request, context -> {request, context} end
    composed = Generic.compose_contextual(handler, [stage])

    assert composed.("request", %{tenant: "t-2"}) ==
             {"request", %{tenant: "t-2", seen: true}}
  end
end
